### ISSUE TITLE
Questions about joint attention and training

### ISSUE THREAD

**Issue by HashWang-null** [lixirui142/UniCon#2]
Thank you for your contributive work. 

I would like to know whether the joint cross-attention proposed in UniCon can be applied to other generative models, such as Stable Diffusion 3. The paper has explicitly mentioned that self-attention controls image content in Stable Diffusion v1-5, but can it also be effective in other paradigms like DiT? 

Additionally, I attempted to reproduce the training of joint attention but found that the loss was unstable. Do you have any insights on this?

**Comment by lixirui142** [comment]
Hi,

We haven't validated UniCon structure to DiT-based models like SD3, but we plan to test it in our future work. I think self-attention should play a similar role in DiT models.

For the second question, do you have a loss curve? It is natural to have a zig-zag loss curve when training because the training noise and timesteps are sampled randomly. It can lead to different loss scales for each step.

**Comment by HashWang-null** [comment]
Thanks for your patience.

One significant concern is that the MM-DiT module in [Stable Diffusion 3 Medium ](https://huggingface.co/stabilityai/stable-diffusion-3-medium) does not seem to explicitly separate self-attention and cross-attention, which hinders the transfer of UniCon to this model. Simply adding a joint cross-attention module in parallel to the MM-DiT Attention would be too expensive. I wonder the appropriate approach to incorporating UniCon control in such models.

As for the second question, I did not use the provided code entirely but instead re-implemented the training code on a custom dataset. I used a simple MSE loss, and the core module is as follows:
```
joint_kwargs = cross_attention_kwargs.pop('joint_kwargs')
attn_output = self.attn1(
    norm_hidden_states,
    encoder_hidden_states=encoder_hidden_states if self.only_cross_attention else None,
    attention_mask=attention_mask,
    **cross_attention_kwargs,
)
x_ids, y_ids = joint_kwargs['x_ids'], joint_kwargs['y_ids']
x_scale, y_scale = joint_kwargs.get('x_scale', 1.0), joint_kwargs.get('y_scale', 1.0)
joint_scale = joint_kwargs.get('joint_scale', 1.0)
input_ids = torch.cat([x_ids, y_ids])
joint_ids = torch.cat([y_ids, x_ids])
input_norm_hidden_states = norm_hidden_states[input_ids]
joint_norm_hidden_states = norm_hidden_states[joint_ids]
joint_attn_output = self.joint_attn(
    input_norm_hidden_states,
    encoder_hidden_states=joint_norm_hidden_states,
    attention_mask=attention_mask,
    **cross_attention_kwargs,
)
joint_output_x, joint_output_y = joint_attn_output.chunk(2, dim=0)
joint_output_x, joint_output_y = self.post_proj(joint_output_x, joint_output_y)
joint_output_x, joint_output_y = (
    joint_output_x * x_scale, 
    joint_output_y * y_scale,
)
joint_output = torch.zeros_like(attn_output)
x_indexes = x_ids.view(-1, 1, 1).expand(-1, *attn_output.shape[1:])
y_indexes = y_ids.view(-1, 1, 1).expand(-1, *attn_output.shape[1:])
joint_output = joint_output.scatter_reduce(dim=0, index=x_indexes, src=joint_output_x, reduce='sum')
joint_output = joint_output.scatter_reduce(dim=0, index=y_indexes, src=joint_output_y, reduce='sum')
attn_output = attn_output + joint_output * joint_scale
```
I trained for approximately 3000 steps, with an actual batch size of 32 per step (using gradient accumulation). The resulting loss curve is that: 
![Image](https://github.com/user-attachments/assets/a5ba456d-adc0-43bf-9608-9501d5dfba90)

The loss curve fluctuates within a very small range, and the final loss is almost the same as at the initial step. However, **what confuses me the most is that the model I obtained works very well:** its output images are well-aligned with the conditions, and the conditional generation performs well. Does this mean that applying joint cross-attention across all self-attention layers is not always necessary, and perhaps some subtractive operations would not affect UniCon’s normal functioning? 

I am very much looking forward to your explanation, as this would further help the community understand the mechanism of UniCon.

**Comment by lixirui142** [comment]
Thanks for your question.

> The loss curve fluctuates within a very small range, and the final loss is almost the same as at the initial step.
We have observed similar loss curves in our training. This is the loss curve of our depth model:

![Image](https://github.com/user-attachments/assets/cfe151c5-94b2-441d-8c5e-fef84696f30c)

I think it is normal that the final loss almost doesn't change. We are fine-tuning a large-scale pretrained diffusion model, which is an excellent noise predictor at the beginning. So I guess it is hard to observe further loss drop in our training. The model can still learn the correlation between paired inputs during the optimization. Therefore, it sounds good that your model works well.

> Does this mean that applying joint cross-attention across all self-attention layers is not always necessary, and perhaps some subtractive operations would not affect UniCon’s normal functioning?

I did not fully capture why the trained model working well leads to this conclusion. But I agree that we may not always need to apply joint cross-attention to all SA layers. In Table 3 of our paper, you can see that our model works well for conditional generation when applied to only UNet decoders (despite a slight performance drop). So you can try dropping some joint cross-attention layers to improve efficiency.

> I wonder the appropriate approach to incorporating UniCon control in such models.

I haven't considered the problem seriously, but we may directly fine-tune the MM-DiT modules for those models without adding new joint cross-attention modules. Just concat all tokens and do joint self-attention in the MM-DiT modules and train separate Q,K,V adapters for tokens from the image and condition.

**Comment by HashWang-null** [comment]
All my questions have been answered. Thanks again for your patience.
